pkg/resources/management.cattle.io/v3/nodedriver/validator.go

```diff
@@ -0,0 +1,163 @@
+}
+
+type k8sAPIImpl struct {
+	nodeCache controllersv3.NodeCache
+	dynamic   *dynamic.Controller
+}
+
+func NewValidator(nodeCache controllersv3.NodeCache, dynamic *dynamic.Controller) admission.ValidatingAdmissionHandler {
+	return &Validator{admitter: admitter{
+		api: &k8sAPIImpl{
+			nodeCache: nodeCache,
+			dynamic:   dynamic,
+		}},
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Update, admissionregistrationv1.Delete}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	return []admissionregistrationv1.ValidatingWebhook{*admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.ClusterScope, v.Operations())}
+}
+
+// Admitters returns the admitter objects used to validate machineconfigs.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+// Admit is the entrypoint for the validator. Admit will return an error if it unable to process the request.
+// If this function is called without NewValidator(..) calls will panic.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	oldObject, newObject, err := objectsv3.NodeDriverOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode object from request: %w", err)
+	}
+
+	// the check to see if the driver is being disabled is either when we're
+	// running a delete operation OR an update operation where the active flag
+	// toggles from true -> false
+	var disablingDriver bool
+	if (request.Operation == admissionv1.Delete && oldObject.Spec.Active) ||
+		(request.Operation == admissionv1.Update && !newObject.Spec.Active && oldObject.Spec.Active) {
+		disablingDriver = true
+	}
+
+	if !disablingDriver {
+		return admission.ResponseAllowed(), nil
+	}
+
+	// check if all node resources have been deleted for both cluster types
+	rke1Deleted, err := a.api.rke1ResourcesDeleted(oldObject)
+	if err != nil {
+		return nil, err
+	}
+	rke2Deleted, err := a.api.rke2ResourcesDeleted(oldObject)
+	if err != nil {
+		return nil, err
+	}
+
+	if !(rke1Deleted && rke2Deleted) {
+		return driverInUse, nil
+	}
+
+	return admission.ResponseAllowed(), nil
+}
+
+// // RKE1
+// this one is a bit more clean since we're just looking at nodes with
+// the <displayname> provider
+func (k *k8sAPIImpl) rke1ResourcesDeleted(driver *v3.NodeDriver) (bool, error) {
+	nodes, err := k.nodeCache.List("", labels.Everything())
+	if err != nil {
+		return true, err
```

Comment:
1. wrap errors from external function calls.
2. return the default value of a bool when err is not nil.
```suggestion
		return false, err
```